Source/com/drew/imaging/png/PngChunkReader.java

```diff
@@ -86,7 +86,7 @@
             // Process the next chunk.
             int chunkDataLength = reader.getInt32();
 
-            if (chunkDataLength < 0)
+            if ((chunkDataLength < 0) || (chunkDataLength > reader.available()))
```

Comment:
I'm afraid that this isn't a valid use of `available()`. Here's the JavaDoc for the method:
> Returns an estimate of the number of bytes that can be read (or skipped over) from this SequentialReader without blocking by the next invocation of a method for this input stream. A single read or skip of this many bytes will not block, but may read or skip fewer bytes. 
>
>Note that while some implementations of SequentialReader like SequentialByteArrayReader will return the total remaining number of bytes in the stream, others will not. It is never correct to use the return value of this method to allocate a buffer intended to hold all data in this stream.
>
>Returns:
an estimate of the number of bytes that can be read (or skipped over) from this SequentialReader without blocking or 0 when it reaches the end of the input stream.

This is a more fundamental "problem", the nature of a stream is that the total size might be unknown. For `SequentialByteArrayReader` using `available()` would yield the correct results, while for `StreamReader` (and potentially other implementations` it would not.

I don't know how a "sane maximum value" could be deducted, but I guess some more finesse is needed.

Reply:
OK, you are right about this. There are implementations of `InputStream` which might return unexpected results for `available()`. I didn't realized that when I implemented my fix. I just thought `available()` is the best chance there is for estimating stream size. And only thought about metadata-extractor as a tool for bounded streams.

What do you suggest then? Using `available()` might fail for some crazy streams. And technically there is nothing wrong with allocating array up to max Integer. It can cause some Exceptions which are impossible to catch/prevent from happening outside the library.

Reply:
I'm not sure I have a suggestion from the top of my head. I guess a check could be made with `instanceof` and then use your logic if the reader is a `SequentialByteArrayReader `, but this still leaves the question of what to do with all the other cases.

Regarding `available()`, it's not just "crazy streams" where this fails. Generally, for `FileInputStream` it returns the remaining bytes of the file, and I think that's the reason why many seem to think it reports "the remainder of the stream". For a network stream for example, `available()` will generally just report the remainder of bytes in the buffer at that time, which will probably be a relatively small value.

Reply:
It is true that this might be unnecessary optimization as allocating large buffer is generally OK and reading more bytes from a stream than it actually contains should result in `java.io.EOFException: End of data reached.`

Let me know though if you come up with any solution or a trick how to handle this issue. I realize it is not caused by the library, but such threat of OOM is really undesirable one.